### ISSUE TITLE
Support the `run` command

### ISSUE THREAD

**Issue by brettcannon** [mamba-org/mamba#1467]
We have started relying on `conda run` in the [Python extension for VS Code](https://marketplace.visualstudio.com/items?itemName=ms-python.python) to support conda appropriately (trying to run an activate conda environment from a subprocess is a bit tough otherwise). We would love to be able to say we support (micro)mamba as well! But based on https://mamba.readthedocs.io/en/latest/user_guide/micromamba.html it looks like the `run` command is not supported by mamba yet.

**Comment by JohanMabille** [comment]
We forgot to update the documentation, but last micromamba (0.21.0) supports the `run` command. The arguments are slightly different from `conda` though.

**Comment by wolfv** [comment]
Actually, for `mamba` the `run` command just forward to `conda run` and should work the same. For `micromamba` we did make some changes to the CLI:

```
Usage: /Users/wolfvollprecht/Programs/mamba/cbuild/./micromamba/micromamba run [OPTIONS]

Options:
  -h,--help                   Print this help message and exit
  -a,--attach TEXT            Attach to stdin, stdout and/or stderr. -a "" for disabling stream redirection
  --cwd TEXT                  Current working directory for command to run in. Defaults to cwd
  -d,--detach                 Detach process from terminal
  --clean-env                 Start with a clean environment
  -e,--env TEXT ...           Add env vars with -e ENVVAR or -e ENVVAR=VALUE


Prefix options:
  -r,--root-prefix TEXT       Path to the root prefix
  -p,--prefix TEXT            Path to the target prefix
  -n,--name TEXT              Name of the target prefix
```

**Comment by brettcannon** [comment]
Thanks for the info! The key thing that micromamba seems to be missing (from our perspective/needs from the conda API) is `--no-capture-output` and `--live-stream`.

Having said that, I have a [proposal for a plug-in system for the Python Launcher for Unix](https://github.com/brettcannon/python-launcher/discussions/168) which could generalize to a mechanism for editors to discover interpreters and environments. So in this case, someone could provide a binary which gave the relevant details on how to execute using micromamba and we wouldn't care what your API was as you would tell VS Code how to execute code instead of us keeping track of us.

If we implemented this idea do you think you would be interested in supporting it?

**Comment by wolfv** [comment]
@brettcannon yes, sure, we'd be down to support it. How are you currently using the flags? `micromamba run` will, by default, attach all streams to the child process. Or you can select which streams to attach using `-a stdin -a stdout` ...

**Comment by wolfv** [comment]
I still don't understand the flags of conda 100% here. 

If I run 

* `conda run --live-stream --no-capture-output -n jupyter jupyter notebook` I seem to get a proper live stream of the output.
* With only `--live-stream` (no `--no-capture-output`) I don't get any output.
* Without any argument I don't get any output for `jupyter notebook`.
* With `conda run --no-capture-output -n jupyter jupyter notebook` I get a live stream.

Anyways I found these flags more confusing so we opted to go with what `docker run` is using.

If you tell me what the expected behavior is and how you're using the flags we can implement them as hidden flags and  ignore them on our side.

**Comment by brettcannon** [comment]
The key flag for us is `--no-capture-output`, but we toss on `--live-stream` just to be safe. If you leave those flags off you already noticed that conda controls stdin/stdout for the Python code it runs and doesn't print it out to conda's stdin/stdout until the process exits. Not good when you're trying to stream results from a linter or you a beginner wondering why `input("my name:")` isn't working. 😉

So if micromamba attaches the stdin/stdout directly from the child process then I suspect the flags are redundant/unnecessary for you.

**Comment by jonashaag** [comment]
Passing `--live-stream` when `--no-capture-output` is passed is a no-op, see https://github.com/conda/conda/blob/20622423ace946475ba1680b74ea99c52ed2230e/conda/gateways/subprocess.py#L87-L93.

`--live-stream` means that stdout/stderr is captured internally in Conda's `subprocess` call _as well as_ live streamed to stdout/stderr. Which in the special case of `conda xyz` it means that this piece of code has `response.stdout/stderr` set to the output of the `conda xyz` implementation https://github.com/conda/conda/blob/35cd47952fe694b49375a72bdbaf7834570fdc2d/conda/cli/main_run.py#L30 ... just to be ignored right after https://github.com/conda/conda/blob/35cd47952fe694b49375a72bdbaf7834570fdc2d/conda/cli/main_run.py#L49.

So, in summary, `--live-stream` doesn't seem to make any sense for `conda xyz` commands. Maybe it's used internally somewhere.

Actually I think I will propose to remove this from Conda

**Comment by jonashaag** [comment]
https://github.com/conda/conda/issues/11209

**Comment by brettcannon** [comment]
@jonashaag thanks for the info! I've opened https://github.com/microsoft/vscode-python/issues/18511 on our end to make sure we are not using the option anywhere.

**Comment by woodgiraffe** [comment]
> @brettcannon yes, sure, we'd be down to support it. How are you currently using the flags? `micromamba run` will, by default, attach all streams to the child process. Or you can select which streams to attach using `-a stdin -a stdout` ...

@wolfv `-a stdin` does not currently seem implemented, see ↓
https://github.com/mamba-org/mamba/blob/master/micromamba/src/run.cpp#L115

I think without it, it's currently impossible to have the `stdin` of the command to be run attached to a `pts`, passing `-a "stdout stderr"` attaches those fds to a `pts` fine tho

`conda` attaches a `pts` to `stdin` via `--no-capture-output`

thanks for the work on micromamba, I've been greatly enjoying the improved ux

**Comment by woodgiraffe** [comment]
> @woodgiraffe I think `stdin` is attached by default (ie. when not passing any `-a`)
> 
> Maybe @JohanMabille can comment more, he implemented this.

Thanks for the reply!

Here is what I tried:
```
# `test -t 0` → checks if fd is opened on a terminal

$ micromamba run -n base /bin/bash -c "test -t 0"
✗ 1

$ micromamba run -n base /bin/bash -c "test -t 1"
✓ 0

$ micromamba run -a "stdin stderr stdout" -n base /bin/bash -c "test -t 0"
✗ 1

$ micromamba run -a "stdin stderr stdout" -n base /bin/bash -c "test -t 1"
✓ 0

$ micromamba run -a "" -n base /bin/bash -c "test -t 0"
? 127

$ micromamba run -a "" -n base /bin/bash -c "test -t 1"
? 127

$ conda run -n base /bin/bash -c "test -t 0"
✗ 1

$ conda run -n base --no-capture-output /bin/bash -c "test -t 0"
✓ 0
```

Host:
```
$ uname -a             → Linux hostname 5.16.5-arch1-1 #1 SMP PREEMPT Tue, 01 Feb 2022 21:42:50 +0000 x86_64 GNU/Linux
$ micromamba --version → 0.21.1
```

**Comment by wolfv** [comment]
Thanks @woodgiraffe for the tests! I think this PR fixes it: https://github.com/mamba-org/mamba/pull/1488

Will land in the next release.

**Comment by jonashaag** [comment]
Can we close this?

**Comment by wolfv** [comment]
Yeah, I think so. Looks like we also have quite a few people using the `run` command these days :)
